File: src/bindings/libbpfcontain.rs
// SPDX-License-Identifier: GPL-2.0-or-later
//
// BPFContain - Container security with eBPF
// Copyright (C) 2020  William Findlay
//
// Dec. 29, 2020  William Findlay  Created this.

//! Function bindings for libbpfcontain.

use anyhow::{bail, Result};

use super::raw;
use crate::policy::Policy;

/// Place the current process into a container with ID `policy_id`.
pub fn containerize(policy: &Policy) -> Result<()> {
    let result = unsafe {
        raw::containerize(
            policy.policy_id(),
            policy.default_taint() as u8,
            policy.default_deny() as u8,
        )
    };

    match result {
        0 => Ok(()),
        n if n == -libc::EAGAIN => bail!("Failed to call into uprobe"),
        n if n == -libc::ENOENT => bail!("No such container with ID {}", policy.policy_id()),
        n if n == -libc::EINVAL => bail!("Process is already containerized or no room in map"),
        n => bail!("Unknown error {}", n),
    }
}

File: src/libbpfcontain/libbpfcontain.c
// SPDX-License-Identifier: GPL-2.0-or-later
//
// BPFContain - Container security with eBPF
// Copyright (C) 2020  William Findlay
//
// Dec. 29, 2020  William Findlay  Created this.

#include "../include/libbpfcontain.h"

#include <errno.h>

/* ========================================================================= *
 * Commands                                                                  *
 * ========================================================================= */

/* Commands should always be declared as follows:
 *
 *    void do_COMMAND(int *ret, ...) { }
 *
 *    int COMMAND(...)
 *    {
 *        int ret = -EAGAIN;
 *
 *        do_start_container(&ret, ...);
 *
 *        if (ret < 0) {
 *            errno = ret;
 *        }
 *
 *        return ret;
 *    }
 *
 */

static void do_containerize(int *ret, u64 policy_id, u8 tainted)
{
}

int containerize(u64 policy_id, u8 tainted)
{
    int ret = -EAGAIN;

    do_containerize(&ret, policy_id, tainted);

    if (ret < 0) {
        errno = -ret;
        return ret;
    }

    return 0;
}
